test/test-tcp-close-after-read-timeout.c

```diff
@@ -0,0 +1,170 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ */
+
+#include "uv.h"
+#include "task.h"
+
+static uv_tcp_t client;
+static uv_tcp_t connection;
+static uv_connect_t connect_req;
+static uv_timer_t timer;
+
+static int read_cb_called;
+static int on_close_called;
+
+static void on_connection(uv_stream_t* server, int status);
+
+static void on_client_connect(uv_connect_t* req, int status);
+static void on_client_alloc(uv_handle_t* handle, size_t suggested_size, uv_buf_t* buf);
+static void on_client_read(uv_stream_t* stream, ssize_t nread, const uv_buf_t* buf);
+static void on_client_timeout(uv_timer_t* handle);
+
+static void on_close(uv_handle_t* handle);
+
+
+static void on_client_connect(uv_connect_t* conn_req, int status) {
+  int r;
+
+  r = uv_read_start((uv_stream_t*)&client, on_client_alloc, on_client_read);
+  ASSERT_EQ(r, 0);
+
+  r = uv_timer_start(&timer, on_client_timeout, 1000, 0);
+  ASSERT_EQ(r, 0);
+}
+
+
+static void on_client_alloc(uv_handle_t* handle, size_t suggested_size, uv_buf_t* buf) {
+  static char slab[8];
+  buf->base = slab;
+  buf->len = sizeof(slab);
+}
+
+
+static void on_client_read(uv_stream_t* stream, ssize_t nread, const uv_buf_t* buf) {
+  ASSERT_LT(nread, 0);
+  read_cb_called++;
+}
+
+
+static void on_client_timeout(uv_timer_t* handle) {
+  ASSERT_EQ(handle, &timer);
+  ASSERT_EQ(read_cb_called, 0);
+  uv_read_stop((uv_stream_t*)&client);
+  uv_close((uv_handle_t*)&client, on_close);
+  uv_close((uv_handle_t*)&timer, on_close);
+}
+
+
+static void on_connection_alloc(uv_handle_t* handle,
+                     size_t suggested_size,
+                     uv_buf_t* buf) {
+  static char slab[8];
+  buf->base = slab;
+  buf->len = sizeof(slab);
+}
+
+
+static void on_connection_read(uv_stream_t* stream, ssize_t nread, const uv_buf_t* buf) {
+  ASSERT_EQ(nread, UV_EOF);
+  read_cb_called++;
+  uv_close((uv_handle_t*)stream, on_close);
+}
+
+
+static void on_connection(uv_stream_t* server, int status) {
+  int r;
+
+  ASSERT_EQ(status, 0);
+  ASSERT_EQ(uv_accept(server, (uv_stream_t*)&connection), 0);
+
+  r = uv_read_start((uv_stream_t*)&connection, on_connection_alloc, on_connection_read);
+  ASSERT_EQ(r, 0);
+}
+
+
+static void on_close(uv_handle_t* handle) {
+  ASSERT(handle == (uv_handle_t*)&client || handle == (uv_handle_t*)&connection || handle == (uv_handle_t*)&timer);
```

Comment:
style nit: wrap this at 80 characters
```suggestion
  ASSERT(handle == (uv_handle_t*) &client || handle == (uv_handle_t*) &connection || handle == (uv_handle_t*) &timer);
```